codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/MiddlewareBuilder.java

```diff
@@ -100,18 +101,65 @@ public void render(RubyCodeWriter writer, GenerationContext context,
         ServiceShape service = context.service();
 
         for (MiddlewareStackStep step : MiddlewareStackStep.values()) {
-            List<Middleware> orderedStepMiddleware = middlewares.get(step)
-                    .stream()
-                    .filter((m) -> m.includeFor(model, service, operation))
-                    .sorted(Comparator.comparing(Middleware::getOrder))
-                    .collect(Collectors.toList());
+            List<Middleware> orderedStepMiddleware = resolveAndFilter(step, model, service, operation);
 
             for (Middleware middleware : orderedStepMiddleware) {
                 middleware.renderAdd(writer, context, operation);
             }
         }
     }
 
+    private List<Middleware> resolveAndFilter(MiddlewareStackStep step, Model model, ServiceShape service,
+                                              OperationShape operation) {
+        Set<Middleware> resolved = new HashSet<>();
+        Map<Middleware, Integer> order = new HashMap<>();
+        Map<String, Middleware> klassToMiddlewareMap = new HashMap<>();
+
+
+        List<Middleware> filteredMiddleware = middlewares.get(step)
+                .stream().filter((m) -> m.includeFor(model, service, operation))
+                .collect(Collectors.toList());
+
+        filteredMiddleware.forEach((m) -> klassToMiddlewareMap.put(m.getKlass(), m));
+
+        try {
+            for (Middleware middleware : filteredMiddleware) {
+                resolve(middleware, resolved, order, klassToMiddlewareMap);
+            }
+        } catch (StackOverflowError e) {
```

Comment:
It would be nice to detect a cycle intelligently without rescuing a stack overflow / out of memory

Reply:
Yeah good idea - pushed a change to detect.